netsim/extra/evpn.multihoming/plugin.py

```diff
@@ -0,0 +1,206 @@
+
+from netsim.utils import log
+from netsim.augment import addressing, devices
+from netsim.modules import _dataplane
+from netsim import api,data
+from box import Box
+import netaddr
+
+_config_name = 'evpn.multihoming'
+_requires = [ 'evpn' ]
+_execute_after = [ 'evpn', 'lag' ]
+
+# Ethernet Segments can be supported on LAG and "physical interfaces", which means the following types for netlab
+_es_supported_on = [ 'lag', # LAG interfaces
+                     'p2p', 'stub', 'lan'  # other physical interfaces
+                    ]
+
+ID_SET = 'esi_auto_id'
+_auto_segments = {}
+
+# structure used to report used ESI, to generate user warnings
+_esi_stats = Box(default_box=True,box_dots=True)
+# count ESI usage references
+_esi_stats.esi_reference = {}
+# map used lacp sys id for ESI
+_esi_stats.esi_lacp_id = {}
+# map assigned LACP ID to node+interface
+_esi_stats.esi_lacp_id_info = {}
+
+# transform a 10-bytes ESI value to int
+#  consider only the first most significant 5 bytes (excluding initial 00) - see below
+def _esi_10bytes_to_int(esi: str) -> int:
+    esi_l = esi.split(":")
+    esi_hex = "".join(esi_l[1:6])
+    return int(esi_hex, 16)
+
+# Collect any user defined ethernet_segments.xxx.id values (as int) and initialize ID generator
+#  we prefer to use as auto generated IDs only 5 bytes, to have unique ES-Import target and LACP System ID (starting with 02),
+#  so we will consider only the first most significant 5 bytes (excluding initial 00)
+def _init_es_id_generator(topology: Box) -> None:
+    _dataplane.create_id_set(ID_SET)
+    used_esi = set()
+    for es, esdata in topology.get('ethernet_segments', {}).items():
+        if 'id' in esdata:
+            used_esi.add(_esi_10bytes_to_int(esdata.id))
+    _dataplane.extend_id_set(ID_SET, used_esi)
+    # we cannot start from 0
+    _dataplane.set_id_counter(ID_SET, 1)
+
+def _generate_esi_and_lag_unique_id() -> tuple:
+    # 00 02 . 5 bytes value taken from id . 00 00 00
+    next_id = _dataplane.get_next_id(ID_SET)
+    esi_id_hex = "0002{:010x}000000".format(next_id)
+    lacp_sys_id = "02{:010x}".format(next_id)
+    return (
+            ':'.join(a+b for a,b in zip(esi_id_hex[::2], esi_id_hex[1::2])),
+            str(netaddr.EUI(lacp_sys_id, dialect = netaddr.mac_unix_expanded))
+        )
+
+def _generate_es_struct(esname: str) -> dict:
+    global _auto_segments
+    unique_ids = _generate_esi_and_lag_unique_id()
+    _auto_segments[esname] = {
+        'id': unique_ids[0],
+        '_lacp_system_id': unique_ids[1],
+    }
+    return _auto_segments[esname]
+
+def _update_esi_stats(node: Box, interface: Box, esi_name: str, esi_data: dict) -> None:
+    if log.debug_active('plugin'):
+        print(f"evpn.multihoming: CALLED ESI STATS {node.name} {interface.ifname} {esi_name} {esi_data}")
+    # initialize struct if required
+    global _esi_stats
+    if esi_name not in _esi_stats.esi_reference: _esi_stats.esi_reference[esi_name] = []
```

Comment:
No need to change the code (comment for future reference only): it might be easier to have the _esi_stats indexed by esi_name, and then every entry be a data structure (= Box) with 'reference','lacp_id' and 'lacp_id_info'